testing/integration/upgrade_rollback_test.go

```diff
@@ -0,0 +1,195 @@
+
+	// rollback should stop the watcher
+	// killTimeout is greater than timeout as the watcher should have been
+	// stopped on its own, and we don't want this test to hide that fact
+	err = upgradetest.WaitForNoWatcher(ctx, 2*time.Minute, 10*time.Second, 3*time.Minute)
+	require.NoError(t, err)
+
+	// now that the watcher has stopped lets ensure that it's still the expected
+	// version, otherwise it's possible that it was rolled back to the original version
+	err = upgradetest.CheckHealthyAndVersion(ctx, startFixture, startVersionInfo.Binary)
+	assert.NoError(t, err)
+}
+
+// TestStandaloneUpgradeRollbackOnRestarts tests the scenario where upgrading to a new version
+// of Agent fails due to the new Agent binary not starting up. It checks that the Agent is
+// rolled back to the previous version.
+func TestStandaloneUpgradeRollbackOnRestarts(t *testing.T) {
+	define.Require(t, define.Requirements{
+		Local: false, // requires Agent installation
+		Sudo:  true,  // requires Agent installation
+	})
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Start at the build version as we want to test the retry
+	// logic that is in the build.
+	startFixture, err := define.NewFixture(t, define.Version())
+	require.NoError(t, err)
+	startVersionInfo, err := startFixture.ExecVersion(ctx)
+	require.NoError(t, err, "failed to get start agent build version info")
+
+	// Upgrade to an old build.
+	upgradeToVersion, err := upgradetest.PreviousMinor(ctx, define.Version())
+	require.NoError(t, err)
+	endFixture, err := atesting.NewFixture(
+		t,
+		upgradeToVersion,
+		atesting.WithFetcher(atesting.ArtifactFetcher()),
+	)
+	require.NoError(t, err)
+
+	t.Logf("Testing Elastic Agent upgrade from %s to %s...", define.Version(), upgradeToVersion)
+
+	// Use the post-upgrade hook to bypass the remainder of the PerformUpgrade
+	// because we want to do our own checks for the rollback.
+	var ErrPostExit = errors.New("post exit")
+	postUpgradeHook := func() error {
+		return ErrPostExit
+	}
+
+	err = upgradetest.PerformUpgrade(
+		ctx, startFixture, endFixture, t,
+		upgradetest.WithPostUpgradeHook(postUpgradeHook))
+	if !errors.Is(err, ErrPostExit) {
+		require.NoError(t, err)
+	}
+
+	// A few seconds after the upgrade, deliberately restart upgraded Agent a
+	// couple of times to simulate Agent crashing.
+	for restartIdx := 0; restartIdx < 3; restartIdx++ {
+		time.Sleep(10 * time.Second)
+		topPath := paths.Top()
+
+		t.Logf("Restarting Agent via service to simulate crashing")
+		err = install.RestartService(topPath)
+		require.NoError(t, err)
+	}
+
+	// wait for the agent to be healthy and back at the start version
+	err = upgradetest.WaitHealthyAndVersion(ctx, startFixture, startVersionInfo.Binary, 10*time.Minute, 10*time.Second, t)
+	if err != nil {
+		// agent never got healthy, but we need to ensure the watcher is stopped before continuing
+		// this kills the watcher instantly and waits for it to be gone before continuing
```

Comment:
Same comment as https://github.com/elastic/elastic-agent/pull/3477/files#r1341596538 about the "continuing" being unclear.